File: bookstore-service-micronaut/src/main/java/example/micronaut/bookstore/api/kafka/producer/BookProducer.java
package example.micronaut.bookstore.api.kafka.producer;

import example.micronaut.bookstore.api.model.Book;
import io.micronaut.configuration.kafka.annotation.KafkaClient;
import io.micronaut.configuration.kafka.annotation.KafkaKey;
import io.micronaut.configuration.kafka.annotation.Topic;
import io.micronaut.context.annotation.Requires;
import io.micronaut.core.util.StringUtils;

@Requires(property = "kafka.enabled", value = StringUtils.TRUE)
@KafkaClient
public interface BookProducer {

    String TOPIC = "example.bookstore.topic.book";

    @Topic(TOPIC)
    void sendBook(@KafkaKey String key, Book book);
}

File: bookstore-service-micronaut-kotlin/src/main/kotlin/com/example/micronaut/bookstore/kafka/producer/BookProducer.kt
package com.example.micronaut.bookstore.kafka.producer

import com.example.micronaut.bookstore.controller.model.Book
import com.example.micronaut.bookstore.kafka.producer.event.KafkaEvent
import io.micronaut.configuration.kafka.annotation.KafkaClient
import io.micronaut.configuration.kafka.annotation.KafkaKey
import io.micronaut.configuration.kafka.annotation.Topic

@KafkaClient
interface BookProducer {

    companion object {
        const val TOPIC = "example.micronaut.bookstore.book"
    }

    @Topic(TOPIC)
    fun sendBook(@KafkaKey key: String, book: KafkaEvent<Book>)
}